File: mobile/lib/features/verification/verification_api.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../core/api_client.dart';
import '../../core/paged_result.dart';
import '../../core/providers.dart';
import 'verification.dart';

/// Every call the verification feature makes to the API. Screens never call the client
/// directly.
class VerificationApi {
  VerificationApi(this._client);

  final ApiClient _client;

  /// GET /api/verifications?status=AwaitingReporterResponse — the checks waiting on an
  /// answer, latest due first (the API's default sort).
  ///
  /// WHOSE CHECKS IS NOT A PARAMETER. The API scopes a Reporter to the checks on reports they
  /// filed, from the token; nothing sent here widens it. The status goes by NAME.
  Future<PagedResult<VerificationListItem>> pending({int page = 1, int pageSize = 20}) async {
    final json = await _client.get(
      '/api/verifications',
      query: {
        'status': VerificationStatuses.awaitingReporterResponse,
        'page': '$page',
        'pageSize': '$pageSize',
      },
    ) as Map<String, dynamic>;

    return PagedResult.fromJson(json, VerificationListItem.fromJson);
  }

  /// GET /api/verifications/{id}. Someone else's check is a 403, told apart from a 404.
  Future<VerificationDetail> detail(int id) async {
    final json = await _client.get('/api/verifications/$id') as Map<String, dynamic>;
    return VerificationDetail.fromJson(json);
  }

  /// POST /api/verifications/{id}/confirm -> 204.
  ///
  /// A YES/NO AND AN OPTIONAL CAPPED COMMENT, ONCE. No reply comes back and there is no
  /// second round: a second answer is a 409. A blank comment is sent as null rather than as
  /// an empty string, so "said nothing" is stored as nothing.
  Future<void> confirm(int id, {required bool fixed, String? comment}) async {
    final trimmed = comment?.trim() ?? '';
    await _client.post(
      '/api/verifications/$id/confirm',
      body: {
        'confirmed': fixed,
        'comment': trimmed.isEmpty ? null : trimmed,
      },
    );
  }
}

final verificationApiProvider = Provider<VerificationApi>((ref) {
  return VerificationApi(ref.watch(apiClientProvider));
});

/// One page of checks waiting on the reporter. autoDispose so coming back to the list reads
/// it fresh — the sweep may have asked something since. Invalidate the family after answering.
final pendingVerificationsProvider =
    FutureProvider.autoDispose.family<PagedResult<VerificationListItem>, int>((ref, page) {
  return ref.watch(verificationApiProvider).pending(page: page);
});

final verificationDetailProvider =
    FutureProvider.autoDispose.family<VerificationDetail, int>((ref, id) {
  return ref.watch(verificationApiProvider).detail(id);
});

File: web/src/features/verification/services/verificationApi.js
/** Every call the verification feature makes to the API lives here — components never fetch. */

/**
 * Mirrors the API's `VerificationStatus` enum, by NAME, in the enum's own order. The API sends
 * and binds enums by name (JsonStringEnumConverter), so no ordinal is hardcoded here and a
 * member inserted into the C# enum cannot shift the client's meaning.
 */
export const VERIFICATION_STATUSES = [
  'Pending',
  'AwaitingReporterResponse',
  'Confirmed',
  'Reopened',
  'Escalated',
  'Expired',
];

/**
 * The VerificationAgent's verdicts — exactly the strings stored in
 * `VerificationCheck.AgentOutcome` and produced by the agent's `VerificationOutcome`.
 *
 * Deliberately NOT a C# enum on the API side: the outcome is the model's OPINION, stored as a
 * string so nothing reads as acting on it. The check's `status` is the reporter's answer, set in
 * C#. A value missing from this list renders as its raw text in a neutral pill, never a guess.
 */
export const AGENT_OUTCOMES = {
  confirm: 'confirm',
  reopen: 'reopen',
  escalate: 'escalate',
};

const AGENT_OUTCOME_LABELS = {
  [AGENT_OUTCOMES.confirm]: 'Confirm — the repair held',
  [AGENT_OUTCOMES.reopen]: 'Reopen — the repair did not hold',
  [AGENT_OUTCOMES.escalate]: 'Escalate — a pattern, not one failed repair',
};

export function agentOutcomeLabel(outcome) {
  return AGENT_OUTCOME_LABELS[outcome] ?? outcome ?? 'Not judged';
}

/**
 * True when the repair did not hold — by the reporter's answer (Reopened, Escalated) or by the
 * agent's reading (reopen, escalate). Either is a fault that has gone back round the loop, so
 * the detail page shows what it looped back to. Chooses a section to render; decides nothing.
 */
export function didNotHold(check) {
  return (
    check.status === 'Reopened' ||
    check.status === 'Escalated' ||
    check.agentOutcome === AGENT_OUTCOMES.reopen ||
    check.agentOutcome === AGENT_OUTCOMES.escalate
  );
}

/**
 * Mirrors the API's `VerificationSort` query enum. The API takes no direction: `DueAt` is latest
 * due first, and `Status` groups by the stored NAME, alphabetically.
 */
export const VERIFICATION_SORTS = {
  DueAt: 'DueAt',
  Status: 'Status',
};

export const DEFAULT_PAGE_SIZE = 10;

/** Turns "AwaitingReporterResponse" into "Awaiting Reporter Response" for display. */
export function enumLabel(value) {
  return String(value ?? '').replace(/([a-z])([A-Z])/g, '$1 $2');
}

/** A timestamp from the API (`DueAt`, `ReporterRespondedAt`) in the reader's local time. */
export function formatDateTime(value) {
  if (!value) return '—';
  const parsed = new Date(value);
  return Number.isNaN(parsed.getTime()) ? '—' : parsed.toLocaleString();
}

/** "Mar 2026" for a trend point. Built from the parts the API sent, never a date string. */
export function monthLabel(year, month) {
  return new Date(year, month - 1, 1).toLocaleDateString(undefined, { month: 'short', year: 'numeric' });
}

/** A PERCENTAGE the API computed (0–100, two decimals) — formatted, never recomputed. */
export function formatPercent(value) {
  if (value === null || value === undefined) return '—';
  const amount = Number(value);
  return Number.isNaN(amount) ? '—' : `${amount.toLocaleString(undefined, { maximumFractionDigits: 2 })}%`;
}

/**
 * Builds the path for GET /api/verifications. Returned as a string rather than fetched here,
 * because the page reads it through useFetch — which owns loading, error and the JWT. Empty
 * filters are left out rather than sent blank.
 *
 * `search` matches the asset tag on the SERVER, across every page. `dateFrom` / `dateTo` are
 * the "YYYY-MM-DD" strings an <input type="date"> produces and go to the API untouched — never
 * through `new Date()`. They bound DueAt, as UTC days with both ends inclusive.
 *
 * WHO SEES WHAT IS NOT A PARAMETER. A Reporter gets the checks on their own reports and a
 * manager gets every check, decided in VerificationService from the token.
 */
export function buildVerificationsPath({
  search = '',
  status = '',
  dateFrom = '',
  dateTo = '',
  sort = VERIFICATION_SORTS.DueAt,
  page = 1,
  pageSize = DEFAULT_PAGE_SIZE,
} = {}) {
  const params = new URLSearchParams();
  params.set('page', String(page));
  params.set('pageSize', String(pageSize));
  params.set('sort', sort);

  if (search.trim()) params.set('search', search.trim());
  if (status) params.set('status', status);
  if (dateFrom) params.set('dateFrom', dateFrom);
  if (dateTo) params.set('dateTo', dateTo);

  return `/api/verifications?${params.toString()}`;
}

/**
 * GET /api/verifications/{id} — the check, the work order's claim, the reporter's answer and
 * the agent's verdict. For a manager, also what has happened to the machine since.
 */
export function buildVerificationPath(id) {
  return `/api/verifications/${encodeURIComponent(id)}`;
}

/**
 * GET /api/analytics/metrics — reopen rate, clarification efficiency and repeat failures.
 * FacilitiesManager and Admin only. The range is optional; the API refuses one that ends
 * before it starts, so the page does not send it.
 */
export function buildMetricsPath({ fromDate = '', toDate = '' } = {}) {
  const params = new URLSearchParams();
  if (fromDate) params.set('fromDate', fromDate);
  if (toDate) params.set('toDate', toDate);
  const query = params.toString();
  return query ? `/api/analytics/metrics?${query}` : '/api/analytics/metrics';
}

/**
 * The reopen-rate trend as chart rows. A month in which nobody answered a check gets a `rate`
 * of null, so the line BREAKS there instead of dropping to 0% — the API sends 0 for "nothing
 * to divide by", and plotted as a point that would read as a month in which every repair held.
 * Nothing is recomputed: `answered` and `reopenRate` are the API's.
 */
export function trendChartRows(monthlyTrend) {
  return (monthlyTrend ?? []).map((month) => ({
    label: monthLabel(month.year, month.month),
    answered: month.answered,
    reopened: month.reopened,
    rate: month.answered > 0 ? Number(month.reopenRate) : null,
  }));
}
